include/glow/Backends/CompiledFunction.h

```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_BACKENDS_COMPILEDFUNCTION_H
+#define GLOW_BACKENDS_COMPILEDFUNCTION_H
+
+namespace glow {
+
+/// Interface containing state necessary to execute a compiled network.
+class CompiledFunction {
```

Comment:
I've got a question thinking about this API.

I've been thinking about the following possible use-case: compile the function once, store the compiled function,  but then load and run it many times. I think this could be useful, as it does not require permanent re-compilations (which may require a lot of time and memory spent on compilations) and allows for caching/storing of compiled functions.

In its current form this interface does not provide any methods to set/change the "execution environment/configuration" of the compiled function. E.g. you cannot provide inputs or outputs and may be set any other kinds of parameters.

I guess it is always the task of the `ExecutionEngine` to set up those execution/configuration parameters, e.g. in its `run` or `runBatch` methods ? If this is the case, we may need to allow for creating of an  `ExecutionEngine` based on a compiled function (i.e. no IR or graph is available).

Reply:
I have a phase 2 in mind for #1227 which separates the compiled function from the runtime state. I haven’t written it up yet.

The idea is to have a `CompiledFunction::bind()` method which returns a third object: a bound function. This function is bound to concrete buffers for inputs and outputs, so multiple threads can use the same compiled function. Each thread would call bind(), and the execute() method would move from the compiled function to he bound function.

Reply:
@stoklund OK, that makes sense! It will make the configuration part much cleaner.

Reply:
I'm confused.  We still have the ability to compile a function once and run multiple times. EE.run copies new input tensors and executes the compiled function; no recompilation is necessary.  Right?

Reply:
@bertmaher I updated #1227 with more details, let's discuss it there.

If I understand correctly, the current scheme in `ExecutionEngine::run()` is not thread safe since data is copied into the variables which are owned by the module.

Reply:
Ah, ok, sorry, I thought the concern was that this diff had broken the ability to run with new data without recompiling.